flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpoint.java

```diff
@@ -302,4 +302,32 @@ public void registerSharedStatesAfterRestored(SharedStateRegistry sharedStateReg
 	public String toString() {
 		return String.format("Checkpoint %d @ %d for %s", checkpointID, timestamp, job);
 	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+
+		CompletedCheckpoint that = (CompletedCheckpoint) o;
+
+		if (checkpointID != that.checkpointID) {
+			return false;
+		}
+		if (timestamp != that.timestamp) {
```

Comment:
If `checkpointID` is truly fulfilling an ID role, it should be the one and only attribute used in comparison and hash code.

Reply:
It should be, so fixing.

Reply:
And there will only be checkpoints of the same JobId in a checkpoint store, so including the JobId should also not be necessary.